design/envoy-shutdown.md

```diff
@@ -0,0 +1,51 @@
+# Envoy Shutdown Probes
+
+Status: _Draft_
+
+This proposal describes how to add better support for identifying the status of Envoy connections during a restart or redeploy of Envoy.
+
+## Goals
+
+- Provide a mechanism to provide feedback on open connections & load inside an Envoy process
+- Allow for a Envoy fleet roll-out minimizing connection loss
+
+## Non Goals
+
+- Guarantee zero connection loss during a roll-out
+
+## Background
+
+The Envoy process, the data path component of Contour, at times needs to be re-deployed.
+This could be due to an upgrade, a change in configuration, or a node-failure forcing a redeployment.
+
+Contour currently implements a `preStop` hook in the container which signals to Envoy to begin draining connections.
+Once Envoy begins draining, the readiness probe on the pod triggers to fail, which in turn causes that instance of Envoy to stop accepting new connections.
+
+The main problem is that the `preStop` hook (which sends the `/healthcheck/fail` request) does not wait until Envoy has drained all connections, so when the container
+is restarted, users can receive errors.
+
+This design looks to add a new component to Contour which will allow for a way to understand if open connections exist in Envoy before sending a `SIGTERM`.
+
+## High-Level Design
+
+Implement a new sub-command of Contour named `envoyshutdown` which will handle sending the healthcheck fail request to Envoy and then begin polling the http/https listeners for active connections from 
+the `/stats` endpoint available on `localhost:9001`.
+
+## Detailed Design
+
+- Implement new subcommand in Contour called `envoyshutdown`
+- This command will expose an http endpoint over a specific port (e.g. `8090`) and path `/shutdown`. 
+- A new container will be added to the Envoy Daemonset pod which will run this new sub-command (e.g. `contour envoyhealthcheck`)
+- This new container as well as the Envoy container will be updated to use an http preStop hook (see example below)
+- When the pod gets a request to shutdown, the `preStop` hook will send a request to `localhost:8090/shutdown` which will tell Envoy to begin draining connections as well as start polling for active connections and will block until that reaches zero
```

Comment:
This will need some configuration options. Some workloads will never drain to 0 or can take hours to drain to low numbers of connections. We should plan to have a configurable maximum drain time, and a connection count threshold.

Reply:
The max time can be configured by the termination grace period, if that time is met, the Kubernetes will forcibly kill everything. 

I don't see a need to add an extra timeout given this one exists currently.

Reply:
Yeh the timeout is arguable, but the connection count threshold is still needed IMHO.

Reply:
sounds good, I added a new parameter